### PR TITLE
Redirect to the last visited page after login

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,7 +22,7 @@ help:
 ## Project setup
 ##---------------------------------------------------------------------------
 
-start: build up app/config/parameters.yml db rabbitmq-fabric web/built assets-amp var/public.key perm  ## Install and start the project
+start: build up app/config/parameters.yml db rabbitmq-fabric web/built assets-amp assets-apps var/public.key perm  ## Install and start the project
```
Comment (illusionOfParadise) [inline Makefile]: To build an ADI react application.

File: features/redirect.feature [hunk 2]
```diff
@@ -22,3 +22,29 @@ Feature:
 
     When I am on "/inscription?foo=bar"
     Then I should be on "/adhesion?foo=bar"
+
+  Scenario: I am reedireected to the last visited page after login
```
Comment (Remg) [inline features/redirect.feature]: `redirected` :eyes:

File: src/Controller/EnMarche/Security/SecurityController.php [hunk 3]
```diff
@@ -28,7 +28,7 @@ class SecurityController extends Controller
      * @Route("/connexion", name="app_user_login")
      * @Method("GET")
      */
-    public function loginAction(): Response
+    public function loginAction(Request $request): Response
```
Comment (Remg) [inline src/Controller/EnMarche/Security/SecurityController.php]: is this new parameter useful ?

File: Makefile (hunk 1 above)
Comment (illusionOfParadise) [reply Makefile]: Undo after discussion.

File: src/EventListener/LastVisitedPageOnGetListener.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Security\Core\Authentication\Token\AnonymousToken;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\Security\Http\FirewallMapInterface;
+
+class LastVisitedPageOnGetListener implements EventSubscriberInterface
+{
+    private $ignoredRouteNames = ['app_user_login', 'logout', 'app_membership_activate', 'asset_url', '_profiler', '_wdt'];
```
Comment (Remg) [inline src/EventListener/LastVisitedPageOnGetListener.php]: you can remove routes that does not belong to the main firewall, don't you ?

Comment (illusionOfParadise) [reply src/EventListener/LastVisitedPageOnGetListener.php]: Done.

File: src/Security/EventListener/LastVisitedPageOnGetListener.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            KernelEvents::RESPONSE => ['onKernelResponse', -1],
+        ];
+    }
+
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        $token = $this->security->getToken();
+
+        // Stop if not MasterRequest or not Response != 200
```
Comment (Remg) [inline src/Security/EventListener/LastVisitedPageOnGetListener.php]: or Response != 200 ?

File: src/Security/EventListener/LastVisitedPageOnGetListener.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        // Stop for API routes or for ignored routes
+        if (
+            0 === mb_strpos($request->getRequestUri(), $this->apiPathPrefix)
+            || \in_array($request->attributes->get('_route'), self::IGNORED_ROUTES)
+        ) {
+            return;
+        }
+
+        if (self::FIREWALL_NAME === $this->getFirewallName($request)) {
```
Comment (Remg) [inline src/Security/EventListener/LastVisitedPageOnGetListener.php]: IMHO this test should be the very first

but it doesnt change much

File: src/Security/EventListener/LastVisitedPageOnGetListener.php (hunk 5 above)
Comment (illusionOfParadise) [reply src/Security/EventListener/LastVisitedPageOnGetListener.php]: Yes, we will save the page as last visited if the response was succeeded, that's why we ignore responses with the status code different from 200.

Comment (ottaviano) [reply src/Security/EventListener/LastVisitedPageOnGetListener.php]: need remove `not` from comment :D

Comment (illusionOfParadise) [reply src/Security/EventListener/LastVisitedPageOnGetListener.php]: ah :D

File: src/Security/EventListener/LastVisitedPageOnGetListener.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            KernelEvents::RESPONSE => ['onKernelResponse', -1],
+        ];
+    }
+
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        $token = $this->security->getToken();
+
+        // Stop if no Session or it's not a anonymous
```
Comment (thomasbisignani) [inline src/Security/EventListener/LastVisitedPageOnGetListener.php]: `an anonymous`

**Pull request merged**